src/tools/rbd_mirror/Replayer.cc

```diff
@@ -0,0 +1,125 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/debug.h"
+#include "common/errno.h"
+#include "include/stringify.h"
+#include "Replayer.h"
+
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd-mirror: "
+
+using std::map;
+using std::string;
+using std::unique_ptr;
+using std::vector;
+
+namespace rbd {
+namespace mirror {
+
+Replayer::Replayer(RadosRef primary_cluster, const peer_t &peer) :
+  m_lock(stringify("rbd::mirror::Replayer ") + stringify(peer)),
+  m_peer(peer),
+  m_primary(primary_cluster)
+{
+}
+
+Replayer::~Replayer()
+{
+}
+
+int Replayer::start()
+{
+  int r = m_remote->init2(m_peer.client_name.c_str(),
+			  m_peer.cluster_name.c_str(), 0);
+  if (r < 0) {
+    derr << "error initializing remote cluster for " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = m_remote->conf_read_file(nullptr);
+  if (r < 0) {
+    derr << "could not read ceph conf for " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = m_remote->connect();
+  if (r < 0) {
+    derr << "error connecting to remote cluster " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  string cluster_uuid;
+  r = m_remote->cluster_fsid(&cluster_uuid);
+  if (r < 0) {
+    derr << "error reading cluster uuid from remote cluster " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  if (cluster_uuid != m_peer.cluster_uuid) {
+    derr << "configured cluster uuid does not match actual cluster uuid. "
+	 << "expected: " << m_peer.cluster_uuid
+	 << " observed: " << cluster_uuid << dendl;
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+void Replayer::stop()
+{
+  m_remote->shutdown();
+}
+
+void Replayer::set_sources(const map<int64_t, set<string> > &images)
```

Comment:
It certainly is less work / less state tracking if we just say that is mirroring is enabled on a pool and if journaling is enabled on an image, it should be mirrored.  But if we need to track deletion status, I don't know if it saves that much effort.

WRT to designating primary / secondary clusters, it certainly can add some additional sanity safety checks.  i.e. prevent image creation, deletion, modification, etc on a secondary cluster pool.  Failover, especially in an automated, split-brain situation becomes a little more troublesome I think, though.  In an OpenStack environment, is there a standard way to initiate a failover to a new data center or are ppl required to roll their own solutions?  

An alternative approach is to track epochs within each image (or associated journal).  When an image acquires the exclusive lock, it starts a new epoch.  For the RBD mirroring daemon, we could have it be the remote journal reader and have it directly re-write the event entries in the local journal (while keeping the associated image open via librbd in a special live replay mode of operation where it reuses the existing replayer logic to apply the journal events).  Another alternative is a new API method to pass an opaque journal event to librbd image for it to directly apply (basically don't want to re-implement replay logic in rbd-mirror).  The mirroring daemon would know it should be pulling data from a peer if the peer has the highest epoch number (technically it should go sequentially between all known peers to copy epoch events in-order).   

We can also keep track of a period to handle the case of an unclean transition between clusters.  On rbd_close, we can record a "safely closed" event in the journal.  If that event is missing when acquiring the exclusive lock and replaying a peer's journal (i.e. after a cluster transition), we know that we need to start a new period.  This in effect would allow the peer to detect that it is now inconsistent with the local copy (and would require intervention by someone to pick the real copy).  This might require some interlocks between the local rbd-mirror and the librbd client to ensure all available remote events have been safely recorded locally.

I like the idea of a special file that can hold all the known per-image errors / state for rbd-mirror.  Much cleaner than attempting to scan the logs for error state and easy to integrate with the storage manager.